src/FieldWarning/Assets/Ingame/UI/InputManager.cs

```diff
@@ -30,388 +32,327 @@ namespace PFW.Ingame.UI
-                break;
-
-            case MouseMode.firePos:
-                ApplyHotkeys();
-
-                if (Input.GetMouseButtonDown(0))
-                    _selectionManager.DispatchFirePosCommand();
-
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift))
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-
-                break;
-
-            case MouseMode.reverseMove:
-                ApplyHotkeys();
-                if (Input.GetMouseButtonDown(0)) {
-                    MoveGhostsToMouse();
-                    _selectionManager.DispatchMoveCommand(
-                            false, MoveWaypoint.MoveMode.reverseMove);
-                }
-
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift))
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-                break;
-
-            case MouseMode.fastMove:
-                ApplyHotkeys();
-                if (Input.GetMouseButtonDown(0)) {
-                    MoveGhostsToMouse();
-                    _selectionManager.DispatchMoveCommand(
-                            false, MoveWaypoint.MoveMode.fastMove);
-                }
+            Normal,
+            Purchasing,
+            FirePosition,
+            ReverseMove,
+            FastMove
+        }
 
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift)) 
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-                break;
+        public Modes MouseState { get; private set; } = Modes.Normal;
 
-            default:
-                throw new Exception("impossible state");
-            }
+        public MatchSession Session
+        {
+            get => this._session;
+            set => this._session = this._session ?? value;
         }
 
-        public void OnGUI()
+        private PlayerData LocalPlayer => this.Session.LocalPlayer.Data;
+
+        public static bool KeyDown(KeyCode keyCode)
         {
-            _selectionManager.OnGUI();
+            return Input.GetKey(keyCode);
         }
 
-        private void ShowGhostUnitsAndMaybePurchase(RaycastHit terrainHover)
+        public void AddSpawnPoint(SpawnPointBehaviour monoBehavior)
         {
-            // Show ghost units under mouse:
-            SpawnPointBehaviour closestSpawn = GetClosestSpawn(terrainHover.point);
-            _currentBuyTransaction.PreviewPurchase(
-                terrainHover.point,
-                2 * terrainHover.point - closestSpawn.transform.position);
+            if (monoBehavior is null)
+                throw new ArgumentNullException(nameof(monoBehavior), "SpawnPointBehavior cannot be null or empty.");
 
-            MaybePurchaseGhostUnits(closestSpawn);
+            if (!this._spawnPointList.Contains(monoBehavior))
+                this._spawnPointList.Add(monoBehavior);
```

Comment:
`monoBehavior` is an unfortunate name because it makes it seem like a unity inbuilt (like `transform`). `spawn`? `spawnPoint`?

Reply:
Agree,maybe rename it to spawnPointBehavior ? Or is that against convention ?